python/lsst/pipe/base/connections.py

```diff
@@ -0,0 +1,389 @@
+                   self.outputSchema defined. Its value is what will be saved by the activator framework.
+    * PrerequisiteInput - An input connection type that defines a `lsst.daf.butler.DatasetType` that must
+                          be present at execution time, but that will not be used during the course of
+                          creating the quantum graph to be executed. These most often are things produced
+                          outside the processing pipeline, such as reference catalogs.
+    * Input - Input `lsst.daf.butler.DatasetType`s that will be used in the run method of a PipelineTask.
+              The name used to declare class attribute must match a function argument name in the run
+              method of a PipelineTask. E.g. If the PipelineTaskConnections defines an Input with the name
+              calexp, then the corresponding signature should be PipelineTask.run(calexp, ...)
+    * Output - A `lsst.daf.butler.DatasetType` that will be produced by an execution of a PipelineTask. The
+              name used to declare the connection must correspond to an attribute of a `Struct` that is
+              returned by a `PipelineTask` run method. E.g. if an output connection is defined with the
+              name measCat, then the corresponding PipelineTask.run method must return Struct(measCat=X,..)
+              where X matches the storageClass type defined on the output connection.
+
+    The process of declaring a PipelineTaskConnection class involves parameters passed in the declaration
+    statement.
+
+    The first parameter is dimensions which is an iterable of strings which defines the unit
+    of processing the run method of a corresponding `PipelineTask` will operate on. These dimensions must
+    match dimensions that exist in the butler registry which will be used in executing the corresponding
+    `PipelineTask`.
+
+    The second parameter is labeled defaultTemplates and is conditionally optional. The name attributes of
+    connections can be specified as python format strings, with named format arguments. If any of the
+    name parameters on connections defined in a `PipelineTaskConnections` class contain a template, then
+    a default template value must be specified in the defaultTemplate argument. This is done by passing
+    a dictionary with keys corresponding to a template identifier, and values corresponding to the value
+    to use as a default when formatting the string. For example if ConnectionClass.calexp.name =
+    '{input}Coadd_calexp' then defaultTemplates = {'input': 'deep'}.
+
+    Once a `PipelineTaskConnections` class is created, it is used in the creation of a
+    `PipelineTaskConfig`. This is further documented in the documentation of `PipelineTaskConfig`. For the
+    purposes of this documentation, the relevant information is that the config class allows configuration
+    of connection names by users when running a pipeline.
```

Comment:
One small example of defining connections class and task config class would be super-helpful here.